base_name_search_improved/models/ir_model.py

```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+# © 2016 Daniel Reis
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import models, fields, api
+from openerp import SUPERUSER_ID
+
+
+class ModelExtended(models.Model):
+    _inherit = 'ir.model'
+
+    name_search_ids = fields.Many2many(
+        'ir.model.fields',
+        string='Name Search Fields')
+
+    def _register_hook(self, cr, ids=None):
+
+        def make_name_search():
+
+            @api.model
+            def name_search(self, name='', args=None,
+                            operator='ilike', limit=100):
+                # Regular name search
+                res = name_search.origin(
+                    self, name=name, args=args, operator=operator, limit=limit)
+
+                allowed_ops = ['ilike', 'like', '=']
+                if not res and operator in allowed_ops and self._rec_name:
+                    # Support a list of fields to search on
+                    model = self.env['ir.model'].search(
+                        [('model', '=', str(self._model))])
+                    other_names = model.name_search_ids.mapped('name')
+                    # Try regular search on each additional search field
+                    for rec_name in other_names:
+                        domain = [(rec_name, operator, name)]
+                        recs = self.search(domain, limit=limit)
+                        if recs:
+                            return recs.name_get()
+                    # Try ordered word search on each of the search fields
+                    for rec_name in [self._rec_name] + other_names:
+                        domain = [(rec_name, operator, name.replace(' ', '%'))]
+                        recs = self.search(domain, limit=limit)
+                        if recs:
+                            return recs.name_get()
+                    # Try unordered word search on each of the search fields
+                    for rec_name in [self._rec_name] + other_names:
+                        domain = [(rec_name, operator, x)
+                                  for x in name.split() if x]
+                        recs = self.search(domain, limit=limit)
```

Comment:
That would have the same effect of just using the last of the three searched (the most relaxed one).
The several layers try to improve the relevance of the results.
If I try "555 777" and there is such a number, I don't need the search to return me a "777 555" matching result.
But if not match to "555 777" was found, the algorithm goes for a more relaxed filer, and tried matching any phone containing "555" and "777", in that order, such as "555 xxx 777".
If it still doesn't find anything, then it tries that again, but ignoring word order, so that "777 xxx 555" would match.

Maybe you could try it with real data and confirm if the results are user friendly.

IMO the layering is needed to produce intuitive results for the user.
The approach goes more relaxed in further tries, so using a single filter would be the same as using the last of the three tries.

If I type "555 777" and there is a phone number containing that string, I expect to find it.


that would proWouldn't work as I intended, and IMO : that would produce the same result as if we serached 

The first one honors spaces: